### ISSUE TITLE
schema_tables.columns refers to used UDTs in a version independent way

### ISSUE THREAD

**Issue by denesb** [scylladb/scylla#9963]
```
cqlsh> CREATE TYPE ks.type1 (f1 int, f2 text);                                                                                                                                                        
cqlsh> create table ks.tt (int pk PRIMARY KEY, v ks.type1); 
cqlsh> select * from system_schema.columns where keyspace_name = 'ks' and table_name = 'tt';

 keyspace_name | table_name | column_name | clustering_order | column_name_bytes | kind          | position | type
---------------+------------+-------------+------------------+-------------------+---------------+----------+-------
            ks |         tt |          pk |             NONE |            0x706b | partition_key |        0 |   int
            ks |         tt |           v |             NONE |              0x76 |       regular |       -1 | type1

(2 rows)
```
As seen above, the `columns` table only has a plain type name. When the type is updated, the UDTs definition is updated both in-memory and on-disk and schemas automatically start referencing the new version. This is fine when all goes well. However if there is a problem in announcing the UDT change, writes that go to a node have no way to determine whether said node has the new UDT definition or not, because writes only carry with them the table schema version -- which is not dependent on the UDT version.

To solve this problem we could mangle the type reference such that it includes a digest (version) of the referenced UDT. We could use a manling scheme which is also human readble, e.g. `$[name}-${version}`, just like we do for table data directories.
When a UDT is updated, this would also trigger an alter on all the dependent table schemas, bumping their versions. With this, writes could detect that a node doesn't have an up-to-date version of the UDT, just based on the table schema version.

**Comment by denesb** [comment]
Related: https://github.com/scylladb/scylla/issues/3760

/cc @tgrabiec @avikivity

**Comment by tgrabiec** [comment]
After c16e4a04239798c41fa140c14eabb250b5422983, the replica should always have at least as recent schema as the coordinator.

**Comment by tgrabiec** [comment]
The problem still exists that when requesting an old UDT schema, replica will use the new UDT schema for responding. Maybe that's also a problem?

**Comment by psarna** [comment]
> After c16e4a0, the replica should always have at least as recent schema as the coordinator.
@tgrabiec if I understood Botond's theory correctly, the problem is that `get_schema_for_write` will not fetch keyspace/udt schema information from other nodes, just table information which seems up to date - because the udt name didn't change.

**Comment by denesb** [comment]
> The problem still exists that when requesting an old UDT schema, replica will use the new UDT schema for responding. Maybe that's also a problem?

Possibly yes. If we had versioned UDT references, each schema would use the correct UDT type object. Currently we overwrite old UDT type with new one in the keyspace metadata object. Old schema objects still alive are not affected, but any new schema object created/deserialized will automatically use the new object after that.

**Comment by denesb** [comment]
> > After [c16e4a0](https://github.com/scylladb/scylla/commit/c16e4a04239798c41fa140c14eabb250b5422983), the replica should always have at least as recent schema as the coordinator.
> 
> @tgrabiec if I understood Botond's theory correctly, the problem is that `get_schema_for_write` will not fetch keyspace/udt schema information from other nodes, just table information which seems up to date - because the udt name didn't change.

Yes, the problem is two-fold: reads/writes will not notice that table schema has changed as an UDT update is not currently reflected in the table schema itself. But even if it did, the schema fetch would not help due to https://github.com/scylladb/scylla/issues/3760.

**Comment by avikivity** [comment]
I don't understand why the UDT name matters. The following should apply:

1. usertype is updated
2. schema version changes
3. new writes have a new schema version
4. everything is blocked until the schema is merged

Types are included in the schema version, see for example do_merge_schema() loading the TYPES table.

**Comment by avikivity** [comment]
Although, I don't see types in schema_mutations:

```c++
// Commutative representation of table schema
// Equality ignores tombstones.
class schema_mutations {
    mutation _columnfamilies;
    mutation _columns;
    mutation_opt _view_virtual_columns;
    mutation_opt _computed_columns;
    mutation_opt _indices;
    mutation_opt _dropped_columns;
    mutation_opt _scylla_tables;
```

**Comment by denesb** [comment]
Exactly: `schema_mutations` stores only the UDT name, the type definition is in the keyspace definition. Updating the type definition currently has *no* impact on the schema version.

**Comment by psarna** [comment]
Maybe when schema_mutations are requested, we should assume the worst (that all types changed) and just always send the keyspace definition as well? It's probably not expected to be very huge.

**Comment by denesb** [comment]
It doesn't help the replica detect that there is a new schema version and that it should update it.

**Comment by tgrabiec** [comment]
I think we should version type definitions, like Botond suggested, so that `schema_mutations` can be used to create the same schema_ptr regardless of what's the current version of the global schema on a given node.

If we don't store the history for the global schema (currently we don't), then we should include type definitions in `schema_mutations`, because the local node may not know the requested version of the type.

The schema pull which we currently have in get_schema_for_write() is not so that schema_ptr sees stuff, it's so that target database objects (tables) have at least as recent schema as that of the request. We could move the pull before the schema_ptr is deserialized from schema_mutations, so that it sees more recent types, but that doesn't solve the problem of using the correct types. It could see the wrong version of the types.

**Comment by tgrabiec** [comment]
Unless, all types have the property that their representation in new version of the type also conforms to the old version of the type without change in value. Then pulling before schema is deserialized would be enough.

**Comment by denesb** [comment]
AFAIK currently you can only add columns to types, not drop them (not sure about alter). So using a newer definition of a type always works for data created with an older type.

**Comment by tgrabiec** [comment]
The question is whether you can also use an older type on data created with a newer type.

**Comment by denesb** [comment]
I don't think that works. What I tried to argue for is that if type definition is backward compatible, then we don't have to keep a full history of type definitions, as long as we can sort versions: a versioned lookup can be considered successful as long as it can find a type definition with version >= $version. This guarantees that as long as you have the correct schema version, you can parse all data for that type, new or old.

**Comment by eliransin** [comment]
> The question is whether you can also use an older type on data created with a newer type.

We have seen cases that it doesn't currently work.
We have seen the following error from one of our users:
`User Defined Type value contained too many fields (expected 5, got 6)`
This suggests that some of the values in the mutation/query are going to be discarded, no?

> AFAIK currently you can only add columns to types, not drop them (not sure about alter). So using a newer definition of a type always works for data created with an older type.

If this is correct, wouldn't the following solve our problem?:
1. include the UDTs mutations in schema digest calculation
2. include the UDTs mutations in every schema pull (following a write mutation/ read request which we don't have the schema for)

It will ensure that a replica has a version of the UDT at least as new as the mutation creator, and that UDT changes are propagated along with schemas.

**Comment by denesb** [comment]
> > The question is whether you can also use an older type on data created with a newer type.
> 
> We have seen cases that it doesn't currently work. We have seen the following error from one of our users: `User Defined Type value contained too many fields (expected 5, got 6)` This suggests that some of the values in the mutation/query are going to be discarded, no?

This means the type definition is older than the data. This is not supported. We only support the type definition being newer then the data. In general our schemas are backward compatible: we don't keep a schema version for all past versions we have data for. This extends to types as well.

> 
> > AFAIK currently you can only add columns to types, not drop them (not sure about alter). So using a newer definition of a type always works for data created with an older type.
> 
> If this is correct, wouldn't the following solve our problem?:
> 
>     1. include the UDTs mutations in schema digest calculation
> 
>     2. include the UDTs mutations in every schema pull (following a write mutation/ read request which we don't have the schema for)
> 
> 
> It will ensure that a replica has a version of the UDT at least as new as the mutation creator, and that UDT changes are propagated along with schemas.

Yes, this would solve the problem. However a keyspace can have many UDTs, each table using only a subset (or none) of them. This would cause a full schema pull to duplicate all these definitions in each table definition.

**Comment by eliransin** [comment]
> Yes, this would solve the problem. However a keyspace can have many UDTs, each table using only a subset (or none) of them. This would cause a full schema pull to duplicate all these definitions in each table definition.

Why this will duplicate the definitions? I think it will only duplicate the need to send them.
We can refine this a bit on the expense of more round trips, the schema pull resoponse can contain the (relevant) UDT versions and the replica will only resolve the pull request once it have pulled all newer UDTs.

**Comment by denesb** [comment]
> > Yes, this would solve the problem. However a keyspace can have many UDTs, each table using only a subset (or none) of them. This would cause a full schema pull to duplicate all these definitions in each table definition.
> 
> Why this will duplicate the definitions? I think it will only duplicate the need to send them. We can refine this a bit on the expense of more round trips, the schema pull resoponse can contain the (relevant) UDT versions and the replica will only resolve the pull request once it have pulled all newer UDTs.

Ah, I misread your suggestion. Including the UDT definition only in schema digest would indeed avoid duplication, but it would make schema mutations non-standalone. You would need external data to calculate its version. With my suggestion (name includes a version), you have all the data you need in the schema mutation itself.

**Comment by eliransin** [comment]
> Ah, I misread your suggestion. Including the UDT definition only in schema digest would indeed avoid duplication, but it would make schema mutations non-standalone. You would need external data to calculate its version. With my suggestion (name

I was only talking about the master digest, not the schema digest.

> includes a version), you have all the data you need in the schema mutation itself.

Just to understand, every time a UDT is changed, the schema for every table that uses it will be updated?
Is sounds to me that you will still need external data (the right UDT).
Moreover, every UDT change will provoke a schema pull (from clients) of all of the tables that uses it (?)

Either way it starting to sound like we need to set up a meeting and brainstorm about this and try to come up with a coherent proposal.
WDYT @denesb @tgrabiec

**Comment by denesb** [comment]
> > Ah, I misread your suggestion. Including the UDT definition only in schema digest would indeed avoid duplication, but it would make schema mutations non-standalone. You would need external data to calculate its version. With my suggestion (name
> 
> I was only talking about the master digest, not the schema digest.


That already includes the keyspace definition and therefore the UDTs as well.

> 
> > includes a version), you have all the data you need in the schema mutation itself.
> 
> Just to understand, every time a UDT is changed, the schema for every table that uses it will be updated?

Yes.

> Is sounds to me that you will still need external data (the right UDT).

Not for schema version calculation, only for schema deserialization. This is the status-quo currently.

>  Moreover, every UDT change will provoke a schema pull (from clients) of all of the tables that uses it (?)

This is very much desired, we want everybody to notice that the schema changed, which is not the case today, when considering only the table schema version (w.r.t. UDT changes).

> Either way it starting to sound like we need to set up a meeting and brainstorm about this and try to come up with a coherent proposal. WDYT @denesb @tgrabiec

I'm fine with a meeting.

**Comment by tgrabiec** [comment]
> > > Ah, I misread your suggestion. Including the UDT definition only in schema digest would indeed avoid duplication, but it would make schema mutations non-standalone. You would need external data to calculate its version. With my suggestion (name
> > 
> > 
> > I was only talking about the master digest, not the schema digest.
> 
> That already includes the keyspace definition and therefore the UDTs as well.
> 
> > > includes a version), you have all the data you need in the schema mutation itself.
> > 
> > 
> > Just to understand, every time a UDT is changed, the schema for every table that uses it will be updated?
> 
> Yes.
> 
> > Is sounds to me that you will still need external data (the right UDT).
> 
> Not for schema version calculation, only for schema deserialization. This is the status-quo currently.
> 
> > Moreover, every UDT change will provoke a schema pull (from clients) of all of the tables that uses it (?)
> 
> This is very much desired, we want everybody to notice that the schema changed, which is not the case today, when considering only the table schema version (w.r.t. UDT changes).

UDT changes also change the schema of tables which depend on them. So they do change table schema version, which is including mutation timestamp in the calculation.

**Comment by denesb** [comment]
> UDT changes also change the schema of tables which depend on them. So they do change table schema version, which is including mutation timestamp in the calculation.

How? Do we rewrite some field in the schema mutations on UDT update to get a later timestamp?

**Comment by psarna** [comment]
> > UDT changes also change the schema of tables which depend on them. So they do change table schema version, which is including mutation timestamp in the calculation.
> 
> How? Do we rewrite some field in the schema mutations on UDT update to get a later timestamp?

Even if we don't, isn't that a nice approximation (easier to write + backward compatible) of your suggestion to append id to the name? Modifications are pretty much guaranteed to come with later timestamps.

**Comment by tgrabiec** [comment]
> How? Do we rewrite some field in the schema mutations on UDT update to get a later timestamp?

type altering statement generates mutations as if the table schema was altered:

https://github.com/scylladb/scylla/blob/master/cql3/statements/alter_type_statement.cc#L100

**Comment by tgrabiec** [comment]
> > > UDT changes also change the schema of tables which depend on them. So they do change table schema version, which is including mutation timestamp in the calculation.
> > 
> > 
> > How? Do we rewrite some field in the schema mutations on UDT update to get a later timestamp?
> 
> Even if we don't, isn't that a nice approximation (easier to write + backward compatible) of your suggestion to append id to the name? Modifications are pretty much guaranteed to come with later timestamps.

We still need to know which version of the type the table uses, and this information is currently absent. So an old table schema may (incorrectly) use a newer type.

**Comment by denesb** [comment]
Using the schema from the first comment:
```CQL
cqlsh> select writetime(type) from system_schema.columns where keyspace_name = 'ks' and table_name = 'tt' and column_name = 'v';

 writetime(type)
------------------
 1643111412798359

(1 rows)
cqlsh> alter TYPE ks.type1 
ADD     RENAME  
cqlsh> alter TYPE ks.type1 ADD f4 int;
cqlsh> select writetime(type) from system_schema.columns where keyspace_name = 'ks' and table_name = 'tt' and column_name = 'v';

 writetime(type)
------------------
 1643721505186815
```
Indeed the write-time is updated, so we do get a schema version bump. But we still have no way to determine whether we have the correct version of the type. Currently the type lookup with the new schema will happily resolve with the old type due to https://github.com/scylladb/scylla/issues/3760. Theoretically, with that issue fixed, we are good. But I still think it has merit to be able to know for sure whether you have the latest and greatest schema, before finding out by parsing exploding in your face.

**Comment by eliransin** [comment]
@denesb I only have one more concern about your proposed solution (which I am going to check).
I would like to validate that no driver is dependent upon the value of the type column.

**Comment by eliransin** [comment]
> @denesb I only have one more concern about your proposed solution (which I am going to check).
> I would like to validate that no driver is dependent upon the value of the type column.

Browsing the code for the cql driver it looks like it is dependent upon.

**Comment by vladzcloudius** [comment]
@denesb @tgrabiec It's not explicitly stated here and I couldn't figure it from the code but is UDT has a (UU)ID at all like other components in the schema?

**Comment by denesb** [comment]
> @denesb @tgrabiec It's not explicitly stated here and I couldn't figure it from the code but is UDT has a (UU)ID at all like other components in the schema?

No, it doesn't AFAIK. UDTs are identified by their name (`$keyspace.$udt_name`).

**Comment by vladzcloudius** [comment]
> > @denesb @tgrabiec It's not explicitly stated here and I couldn't figure it from the code but is UDT has a (UU)ID at all like other components in the schema?
> 
> No, it doesn't AFAIK. UDTs are identified by their name (`$keyspace.$udt_name`).

Well, that's a huge flaw in design. I believe the rest of the issues are just a consequences of this. 
UDT is not different from any other schema component and has to have an ID attached that changes every time its schema changes. 

And since UDT has its own metadata on the client side (AFAIU) then it has to have its own ID too (orthogonal to a KS/CF version).

Otherwise we will keep on chasing all kinds of races during the schema change.
